src/shared/lib/observables/uploadProgressObservable.ts

```diff
@@ -1,95 +1,108 @@
-import { CommonObservable } from '../utils';
-
-type SecondLevelStep = 'upload_files' | 'encrypt_answers' | 'upload_answers';
+import { CommonObservable, wait } from '../utils';
+
+type SecondLevelStep =
+  | 'upload_files'
+  | 'encrypt_answers'
+  | 'upload_answers'
+  | 'completed';
+
+export type UploadProgress = {
+  totalActivities: number | null;
+  currentActivity: number | null;
+  currentActivityName: string | null;
+  totalFilesInActivity: number | null;
+  currentFile: number | null;
+  currentSecondLevelStepKey: SecondLevelStep | null;
+};
 
 export interface IUploadProgressObservableSetters {
-  set totalActivities(value: number);
-  set currentActivity(value: number);
-  set currentActivityName(value: string);
-  set totalFilesInActivity(value: number);
-  set currentFile(value: number);
-  set currentSecondLevelStep(value: SecondLevelStep | null);
+  set totalActivities(value: number | null);
+  set currentActivity(value: number | null);
+  set currentActivityName(value: string | null);
+  set totalFilesInActivity(value: number | null);
+  set currentFile(value: number | null);
+  set currentSecondLevelStepKey(value: SecondLevelStep | null);
+  delay(ms: number): Promise<void>;
   reset(): void;
 }
 
 class UploadProgressObservable
   extends CommonObservable
   implements IUploadProgressObservableSetters
 {
-  private _totalActivities: number = -1;
-
-  private _currentActivity: number = -1; // zero-based
-
-  private _currentActivityName: string = '';
-
-  private _totalFilesInActivity: number = -1;
-
-  private _currentFile: number = -1; // zero-based
-
-  private _currentSecondLevelStep: SecondLevelStep | null = null;
+  private uploadProgress: UploadProgress;
 
   constructor() {
     super();
+    this.uploadProgress = {} as UploadProgress;
```

Comment:
Please wrap `UploadProgress` into Partial<> generic as `this.uploadProgress` initialized with an empty object literal. 
Or you can initialize `this.uploadProgress` with the default values of the `UploadProgress`

Reply:
There's this.reset() call below this line where default values set. Partial is not suitable as we'd miss more strict types check in the consumers of this object.